firstboot: Include <crypt.h> for declaration of crypt() if needed

besser82 opened the pull request:
This is needed on systems, which do not provide a `crypt()` function in their stdlibc.

yuwata commented:
Hmm? What library do you use?

besser82 commented:
I'm using glibc (with libcrypt removed) and libxcrypt (for providing the `crypt()` function) on Fedora 28 / Rawhide.  See:  https://fedoraproject.org/wiki/Changes/Replace_glibc_libcrypt_with_libxcrypt

yuwata commented:
Thanks.
Should we check the existence of libcrypt or libxcrypt in `meson.build`?

yuwata commented:
Yeah, please also provide meson.build update. Also, please mention about libxcrypt in the commit message.

yuwata commented:
Sorry, meson update may not be necessary. lgtm now.

besser82 commented:
I've updated / enhanced the commit message in the rebased commit.

yuwata commented:
Thanks.

poettering commented:
hmm, is there any documentation available for this glibc change, and what downstreams are supposed to do? the feature page from fedora is very poor on details...

why is there no meson change necessary? does glibc redirect the symbols? if it does so, why doesn't it also redirect the header includes? even if glibc redirects the symbols, should we not link directly against the real .so instead?

please elaborate on the details about this all? (ideally in some short blurb in the fedora feature, i.e. "what does this mean for upstreams?" since I doubt this needed change is specific to systemd).

Also, please add a brief comment to the patched lines in the sources too, so that it is clear why we need this ifdeffery here...

keszybz commented:
We need this quickly, so I went ahead and force-pushed an updated version into your branch with the requested comment.

> should we not link directly against the real .so instead?

That's a good question. But without the patch systmed does not build in rawhide, so let's apply it, and figure out any additional changes later.

besser82 commented:
@poettering @keszybz 

I've added a check whether `crypt.h` being present in `meson.build` and changed the conditional in `firstboot.c` to include it, if it is present.  I've extended the comment there as well.

besser82 commented:
`mariadb`, `mysql`, `postgresql`, `perl` and some other packages do it the same way since ages.

besser82 commented:
> hmm, is there any documentation available for this glibc change, and what downstreams are supposed to do? the feature page from fedora is very poor on details...

Well, there is not much documentation about this, because the needed changes (if any) are really simple.


> why is there no meson change necessary? does glibc redirect the symbols? if it does so, why doesn't it also redirect the header includes? even if glibc redirects the symbols, should we not link directly against the real .so instead?

libxcrypt directly replaces `libcrypt.so` from glibc in a binary compatible way, that's why there is no deep change to the build system needed.  If sth. already links to this library (which is mandatory, if one uses the `crypt()` function with glibc) everything is already set.


> please elaborate on the details about this all? (ideally in some short blurb in the fedora feature, i.e. "what does this mean for upstreams?" since I doubt this needed change is specific to systemd).

Will do so during today, though there are really few packages in need for any change; most of them already have been changed upstream.

keszybz commented:
OK, that version is nice too.

poettering commented:
> libxcrypt directly replaces libcrypt.so from glibc in a binary compatible way, that's why there is no deep change to the build system needed. If sth. already links to this library (which is mandatory, if one uses the crypt() function with glibc) everything is already set.

hmm, so let me get this right, the project is called "libxcrypt", but the .so it builds is called "libcrypt.so"? Always and undcnditionally (or only on Fedora?)? That's confusing! could you at least clarify that explicitly in the fedora feature page?

besser82 commented:
@poettering 

I've made the changes to the feature page.

* [The need for `<crypt.h>`](https://fedoraproject.org/wiki/Changes/Replace_glibc_libcrypt_with_libxcrypt#Scope)
* [Notes on libcrypt.so.1 by libxcrypt](https://fedoraproject.org/wiki/Changes/Replace_glibc_libcrypt_with_libxcrypt#Upgrade.2Fcompatibility_impact)

keszybz commented:
artful-s390x and Fedora rawhide and semaphore pass, the rest is being slow today.
